WalletWasabi/Wallets/FilterProcessor/ParallelBlockDownloadService.cs

```diff
@@ -0,0 +1,190 @@
+using Microsoft.Extensions.Hosting;
+using NBitcoin;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Threading;
+using System.Threading.Tasks;
+using WalletWasabi.Logging;
+
+namespace WalletWasabi.Wallets.FilterProcessor;
+
+/// <summary>
+/// Service that opportunistically downloads blocks upfront and in parallel.
+/// </summary>
+public class ParallelBlockDownloadService : BackgroundService
+{
+	/// <summary>Maximum number of parallel block-downloading tasks.</summary>
+	public const int MaxParallelTasks = 5;
+
+	/// <summary>Maximum number of attempts to download a block. If it fails, we drop the block download request altogether.</summary>
+	public const int MaxFailedAttempts = 3;
```

Comment:
I'm not sure about this. The `SmartBlockProvider` is supposed to give blocks or to throw. It makes no sense to retry if the SmartBlockProvider is used, which will always be the case "in real life"

Reply:
Probably, but therefore we must delete the `while true` in P2PBlockProvider, and simply return `null` if we failed. But in this case, retrying 3 times is not enough, and we should retry indefinitely until the reorg is detected by the wallet and the `Remove` is called. The reason is that a node not having a block is quite frequent.

Probably the design would be better that way. In any case, failing to get a block is a black swan event, the wallet need to crash in that case, or at least synchronization can't simply skip it.